infra/README.md

```diff
@@ -0,0 +1,113 @@
+This is a work in progress to use Terraform to manage our cloud
+infrastructure.
+
+Instructions:
+
+- You will need a GCP project.  We assume `gcloud` is configured to
+  point at your project.
+
+- Create a service account for Terraform with Owner role, create a
+  service account key and place it in
+  `$HOME/.hail/terraform_sa_key.json`.
+
+- Enable the the GCP services needed by Hail:
+
+   ```
+   gcloud services enable \
+       compute.googleapis.com \
+       cloudresourcemanager.googleapis.com \
+       servicenetworking.googleapis.com \
+       sqladmin.googleapis.com \
+       container.googleapis.com \
+       serviceusage.googleapis.com
+   ```
+
+- Install terraform.
+
+- Create `$HOME/.hail/global.tfvars` that looks like:
+
+   ```
+   gcp_project = "<gcp-project>"
+   gcp_region = "<gcp-region>"
+   gcp_zone = "<gcp-zone>"
+   domain = "<domain>"
+   ```
+
+- Run `terraform init`.
+
+- Run `terraform apply -var-file="$HOME/.hail/global.tfvars"`.
+  Terraform has created a GKE cluster named `vdc`.  We assume
+  `kubectl` is configured to point at this cluster.
+
+- Go to the Google Cloud console, VPC networks > internal > Private
+  service connection > Private connections to services, and enable
+  Export custom routes to both connections.
+
+You can now install Hail:
+
+- Run `$HAIL/docker/third-party/copy_images.sh`.  This copies some
+  base images from Dockerhub (which now has rate limits) to GCR.
+
+- Generate TLS certificates.  See ../dev-docs/tls-cookbook.md.
+
+- Update $HAIL/config.mk with your infrastructure settings.  You can
+  get settings from the default/global-config secret.
+
+- Run `kubectl -n default apply -f $HAIL/ci/bootstrap.yaml`.
+
+- Build the CI utils image.  Run `make push-ci-utils` in $HAIL/ci.
```

Comment:
This failed without running `make service-base` in $HAIL/docker first.

Reply:
> This failed without running make service-base in $HAIL/docker first.

There should be dependencies for this, I don't see  what's wrong.  push-ci-utils depends on build-ci-utils, and:

```
build-ci-utils:
	$(MAKE) -C ../docker build
	...
```

and in docker/Makefile:

```
build: base-stmp service-base hail-ubuntu
```

> When running this on a Ubuntu VM, I also encountered an odd issue in that Makefile, when stat is called: [ "$(shell bash stat-permissions.sh Dockerfile.base)" = "644" ]. The mode of those files was 664 for me, so I had to run a chmod g-w to get the build working.

Yeah, this is a check for docker caching.  Docker includes files permissions in the layer hash, so unless the permissions agree everywhere, you get painful docker rebuilds.  I added a not about seeing the umask on the VM.  I added a step to set the umask on the VM.